Python: Convert string to boolean if the binding variable is Boolean

pritampan opened the pull request:
This is the equivalent of the [Legacy Python PR ](https://github.com/apache/iceberg/pull/6843).

Fokko commented on python/pyiceberg/expressions/literals.py:
```diff
@@ -562,6 +562,13 @@ def _(self, type_var: DecimalType) -> Literal[Decimal]:
+    @to.register(BooleanType)
+    def _(self, type_var: BooleanType) -> Literal[bool]:
+        if self.value.upper() in ['TRUE', 'FALSE']:
+            return BooleanLiteral(True if self.value.upper() == 'TRUE' else False)
```
```suggestion
            return BooleanLiteral(self.value.upper() == 'TRUE')
```

Fokko commented on python/pyiceberg/expressions/literals.py:
```diff
@@ -562,6 +562,13 @@ def _(self, type_var: DecimalType) -> Literal[Decimal]:
 
+    @to.register(BooleanType)
+    def _(self, type_var: BooleanType) -> Literal[bool]:
+        if self.value.upper() in ['TRUE', 'FALSE']:
```
Can we introduce a variable `value_upper` so we call `.upper()` just once?

pritampan replied:
thanks for your suggestions, made the changes.

Fokko reviewed (commented):
Thanks for opening this PR, and I think this is a great addition. Could you rebase against the latest master? Today we had a failing master.

pritampan commented:
> Thanks for opening this PR, and I think this is a great addition. Could you rebase against the latest master? Today we had a failing master.

Thanks for reviewing, I have rebased this branch and addressed the comments.

Fokko commented:
Thanks @pritampan for opening this PR and @puchengy for the review! 👏🏻
